Change font size

kinow opened the pull request:
These changes close #141

There is a TODO about #335, but for now it's storing the new font size in the `LocalStorage` (i.e. client-side only).

Under the user profile view, now there is a "Preferences" section, with the Font Size.

- Vuetify appears to be setting the fixed font-size only on `<html>` element, and computing the sizes of other elements relative to that value (rem).
- I've adjusted our code to do the same, except by the Graph component that has some fixed values that I don't dare to change as it would probably cascade the amount of changes necessary and I am not sure if these could be easily fixed
- Reset sets the value from Vuetify sass variable back to the `<html>` element
- The button to increase font size does so with a ratio of 20% for every click, and stores the new value in the `LocalStorage`
- The button to decrease does so with a 20% ratio too (decay), and also stores new value in the `LocalStorage`
- When the application is loaded, it takes care of setting the `<html>` element font size back to what is in `LocalStorage`. If there is nothing in `LocalStorage`, then it does nothing and the sass variable value is applied as before

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.

Will add changelog and tests (unit + e2e) later.

kinow commented:
![image](https://user-images.githubusercontent.com/304786/79709354-14d14680-8316-11ea-90fc-56d6f212188d.png)

![image](https://user-images.githubusercontent.com/304786/79709389-2d416100-8316-11ea-90ac-aecccde1fa40.png)

(There could still be some elements where the change is not applied, e.g. "This is a read-only view of your user" text at the table top seems to be using fixed size. I'll try to tackle some more as I find, but we can also push it to the backlog and fix as we find them I think.)

kinow commented on src/utils/font-size.js:
```diff
@@ -0,0 +1,80 @@
+import styles from '@/styles/index.scss'
+// TODO: const styles = { fontRootSize: '16px' }
```
This is a blocker at the moment. Interestingly, it works with Vue + webpack. You can import `.js` and `.scss` files with no issues. Except that if you run this file with Cypress in the e2e tests, then it complains that it cannot file any tests in your file. With the commented code I was able to get it to run, but it would be preferable to get the value from Vuetify, in case that changes in future releases.

kinow commented on tests/e2e/specs/userprofile.js:
```diff
@@ -9,4 +16,63 @@ describe('User Profile', () => {
+    expect(parseFloat(INITIAL_FONT_SIZE)).to.be.equal(getCurrentFontSize())
+    const clicks = 3
+    // TODO: https://github.com/cypress-io/cypress/issues/695#issuecomment-379817133
+    // Also had to use Promises in order to locate right element with Cypress
```
Note to self: double check

kinow replied:
The promise approach actually fixed it, so we don't need it any more. But good to know about this hack, as it could be useful later (p.s. comparing with Selenium, Cypress needs very little hacking)

kinow commented on tests/unit/utils/font-size.spec.js:
```diff
@@ -0,0 +1,62 @@
+    getItem () { return '{}' },
+    setItem () {}
+  }
+}
```
No `LocalStorage` with JSDOM, interesting too.

kinow commented:
Cypress was skipping the newly added tests. In `userprofile.js` spec, it used one of our JS modules that imported Vuetify's `scss` file (in JS, not css). That resulted in no errors in Cypress GUI or terminal, and Cypress reporting no tests found.

After searching GitHub and StackOverflow, found some people mentioning things like turning off optimization. Tried it, and that fixed the issue. Now tests being executed. Ran the whole e2e tests locally, everything worked fine.

The issue was similar to this one: https://github.com/cypress-io/cypress/issues/6159

oliver-sanders commented:
Looks good, session storage is a good stepping-stone.

I've part-tested, need a more functional setup to go further. Looks like one element isn't responding, I guess this is because the content isn't wrapped in `<p/>`, `<span/>` or whatever:

<img width="384" alt="Screenshot 2020-04-23 at 00 00 27" src="https://user-images.githubusercontent.com/16705946/80041902-8ddac480-84f5-11ea-8996-7e6749b824e8.png">

kinow commented:
> Looks good, session storage is a good stepping-stone.
> 
> I've part-tested, need a more functional setup to go further. Looks like one element isn't responding, I guess this is because the content isn't wrapped in `<p/>`, `<span/>` or whatever:
> 
> <img alt="Screenshot 2020-04-23 at 00 00 27" width="384" src="https://user-images.githubusercontent.com/16705946/80041902-8ddac480-84f5-11ea-8996-7e6749b824e8.png">

Ah, good idea! I was thinking how to tackle that dangling piece of text that wouldn't resize.

First I wrapped it in a paragraph. But then the font-size applied was the `16px` from the vuetify variable (guess it's [applied not only to `html` element](https://github.com/vuetifyjs/vuetify/blob/4140af4c776aa8cb2dbd52ed7a05c44268094ec8/packages/vuetify/src/components/VAlert/VAlert.sass#L13)).

But then I remembered Vuetify also provides classes for each font-size available, and it's documented in their [typography section](https://v15.vuetifyjs.com/en/framework/typography); and `body-1` is the default size, but not fixed as `16px`, but instead set as `1rem !important` (that's what we get in the computed style).

Pushed one more commit that wraps every piece of text that is not already in a `h3` or another element that defines the font-size, using the `body-1` class. Now it appears to be working OK.

Thanks for the testing and good suggestion!

kinow commented:
>Might want to squash some commits before merging to keep our commit history a bit more readable.

Rebased, and moved + squashed some commits together.

hjoliver commented:
@kinow - I'll it to you to deconflict and merge.

kinow commented:
> @kinow - I'll it to you to deconflict and merge.

Sure, thank you Hilary!

kinow commented:
Rebased on `master`. Also did a quick smoke test using offline mode, no issues found.
